### PR TITLE
Add message in case it's already up-to-date

### PR DESCRIPTION
**Pull request by GuillaumeGomez**

### CODE REVIEW

**Comment by rbtcollins** [comment]
Hmm, can I ask why?
The current output is
```
$ rustup self update
info: checking for self-updates
```
And this combined with the exit code of 0 is pretty unambiguous. If an error occurs, its logged, if there is work to be done it is logged. I'm not against this per se, but not really for it either.

**Comment by GuillaumeGomez** [comment]
Unambiguous but pretty weird. At first I wondered if it crashed somehow. At least this is now explicit, no need to check the exit code.

**Comment by rbtcollins** [comment]
Well the thing is its also used in the large `rustup update` flow implicitly, and that looks like this:

```
info: installing component 'rust-docs'
 12.0 MiB /  12.0 MiB (100 %)   7.3 MiB/s in  1s ETA:  0s
info: installing component 'rust-std' for 'wasm32-unknown-unknown'
info: checking for self-updates

   stable-x86_64-unknown-linux-gnu updated - rustc 1.39.0 (4560ea788 2019-11-04)
  nightly-x86_64-unknown-linux-gnu updated - rustc 1.41.0-nightly (76a252ea9 2019-12-09)

```

And when there is nothing they show like so:

```
$ rustup update
info: syncing channel updates for 'stable-x86_64-unknown-linux-gnu'
info: syncing channel updates for 'nightly-x86_64-unknown-linux-gnu'
info: checking for self-updates

   stable-x86_64-unknown-linux-gnu unchanged - rustc 1.39.0 (4560ea788 2019-11-04)
  nightly-x86_64-unknown-linux-gnu unchanged - rustc 1.41.0-nightly (76a252ea9 2019-12-09)

```

I don't want to get in the way here, but I wonder if what you really want is to have rustup self updates reported (or not) somehow more consistently with the toolchain information.

If you look at what we output for toolchains, just the one message is consistent, but we then do a summary at the bottom, which we don't do for rustup itself. Perhaps showing something like this would be good:

```
$ rustup update
info: syncing channel updates for 'stable-x86_64-unknown-linux-gnu'
info: syncing channel updates for 'nightly-x86_64-unknown-linux-gnu'
info: checking for self-updates

   stable-x86_64-unknown-linux-gnu unchanged - rustc 1.39.0 (4560ea788 2019-11-04)
  nightly-x86_64-unknown-linux-gnu unchanged - rustc 1.41.0-nightly (76a252ea9 2019-12-09)
  rustup unchanged - rustup 1.20 (2019-10-21)
```

This would give more information

**Comment by GuillaumeGomez** [comment]
Hum... I'm shared on this. I really like this output for example:

```text
$ rustup update
info: syncing channel updates for 'stable-x86_64-unknown-linux-gnu'
info: syncing channel updates for 'nightly-x86_64-unknown-linux-gnu'
info: checking for self-updates
info: already up-to-date
```

Having more "state messages" would be nice. Something like:


```text
$ rustup update
info: syncing channel updates for 'stable-x86_64-unknown-linux-gnu'
info: done
info: syncing channel updates for 'nightly-x86_64-unknown-linux-gnu'
info: done
info: checking for self-updates
info: already up-to-date
```

**Comment by kinnison** [comment]
I quite like @rbtcollins's point that it'd be nice if the format mirrored how we report updates to toolchain channels.  This would require a little more plumbing but would be consistent which users may appreciate.

**Comment by GuillaumeGomez** [comment]
I'll do it in the next days then.

**Comment by kinnison** [comment]
@GuillaumeGomez Happy new year.  I wonder if you might have time in the coming week or so to make progress on this?

**Comment by GuillaumeGomez** [comment]
I'll try to!

**Comment by GuillaumeGomez** [comment]
As suggested, I re-used the code from common to "show".

**Comment by GuillaumeGomez** [comment]
Updated!

File: src/errors.rs [hunk 1]
```diff
@@ -362,6 +362,10 @@ error_chain! {
         BrokenPartialFile {
             description("partially downloaded file may have been damaged and was removed, please try again")
         }
+        Generic(v: String) {
+            description("generic error")
+            display("encountered error: {}", v)
+        }
```
Comment (kinnison) [inline src/errors.rs]: I think this change is a bogon.  I can't find use of this anywhere.

File: src/cli/self_update.rs [hunk 2]
```diff
@@ -1435,21 +1436,24 @@ pub fn update() -> Result<()> {
...
-    if let Some(ref p) = setup_path {
-        let version = match get_new_rustup_version(p) {
-            Some(new_version) => parse_new_rustup_version(new_version),
-            None => {
-                err!("failed to get rustup version");
-                process::exit(1);
-            }
-        };
+    match prepare_update()? {
+        Some(ref setup_path) => {
```
Comment (kinnison) [inline src/cli/self_update.rs]: Does this need to remain `Some(ref setup_path)` now, or could it be `Some(setup_path)` ?

**Review by kinnison — CHANGES_REQUESTED** [review]
One bogon, one query (not important if you choose to decline it), and then the commits need rebasing because we don't like `fmt` type commits, and also something has changed underneath you so I can't merge now even if I wanted to.

File: src/errors.rs (hunk 1 above)
Comment (GuillaumeGomez) [reply src/errors.rs]: Seems so. Removing it!

File: src/cli/self_update.rs (hunk 2 above)
Comment (GuillaumeGomez) [reply src/cli/self_update.rs]: To be confirmed. I check and update if needed.

**Comment by GuillaumeGomez** [comment]
Updated!

**Review by kinnison — APPROVED** [review]
LGTM. :+1:

**Comment by GuillaumeGomez** [comment]
\o/

**Pull request merged**